### PR TITLE
Add gzip compression option for exporter

### PR DESCRIPTION
**Pull request by kw7oe**
Address #123. WIP, just PR to get early feedback.

**Todo:**

- [x] Add test to verify http exporter with compression enabled
- [x] Implement gzip compression for grpc.
- [x] Add test to verify grpc exporter with compression enabled

### CODE REVIEW

File: apps/opentelemetry_exporter/src/opentelemetry_exporter.erl [hunk 1]
```diff
@@ -135,17 +142,23 @@ export(_Tab, _Resource, #state{protocol=http_json}) ->
...
+    {NewHeaders, NewProto} = case Compression of
+      gzip -> {[{"content-encoding", "gzip"} | Headers], zlib:gzip(Proto)};
+        _ -> {Headers, Proto}
+    end,
     Address = uri_string:normalize(#{scheme => Scheme,
                                      host => Host,
                                      port => Port,
                                      path => Path}),
-    case httpc:request(post, {Address, Headers, "application/x-protobuf", Proto}, [], []) of
+    case httpc:request(post, {Address, NewHeaders, "application/x-protobuf",
```
Comment (kw7oe) [inline apps/opentelemetry_exporter/src/opentelemetry_exporter.erl:160]: my editor autoformat, please let me know what are the recommended formatting I should adopt

File: apps/opentelemetry_exporter/src/opentelemetry_exporter.erl [hunk 2]
```diff
@@ -135,17 +142,23 @@ export(_Tab, _Resource, #state{protocol=http_json}) ->
...
                              endpoints=[#{scheme := Scheme,
                                           host := Host,
                                           port := Port,
                                           path := Path} | _]}) ->
     Proto = opentelemetry_exporter_trace_service_pb:encode_msg(tab_to_proto(Tab, Resource),
                                                                export_trace_service_request),
+    {NewHeaders, NewProto} = case Compression of
+      gzip -> {[{"content-encoding", "gzip"} | Headers], zlib:gzip(Proto)};
+        _ -> {Headers, Proto}
+    end,
```
Comment (kw7oe) [inline apps/opentelemetry_exporter/src/opentelemetry_exporter.erl:155]: currently, haven't write any test for this. What I have in mind is reuse `verify_export` in the test suite and configure the state according to the config, so it looks something like this:

```erlang
groups() ->
    [{functional, [], [configuration, span_round_trip, ets_instrumentation_info]},
     {grpc, [], [verify_export]},
     {http_protobuf, [], [verify_export]},
     {http_protobuf, [compression: gzip], [verify_export]}],
```

and:

```erlang
verify_export(Config) ->
    ...
    Compression = ?config(compression, Config),
    {ok, State} = opentelemetry_exporter:init(#{protocol => Protocol,
                                                compression => Compression,
                                                endpoints => [{http, "localhost", Port, []}]}),

```

Is this the right way to do it?

**Comment by kw7oe** [comment]
For passing `gzip` into `grpcbox` `encoding` configuration, I did some research and came across [`grpcbox README`](https://github.com/tsloughter/grpcbox#defining-channels). Given that, I'm still not really sure how to pass it. 

I'm thinking is either here:

```erlang
init(Opts) ->
    %% ...
    Compression = maps:get(compression, Opts1, undefined),
    case maps:get(protocol, Opts1, http_protobuf) of
        grpc ->
            %% Pass in here?
            ChannelOpts = maps:get(channel_opts, Opts1, #{}),
            case grpcbox_channel:start_link(?MODULE, grpcbox_endpoints(Endpoints), ChannelOpts) of
```

or here:

```erlang
export(Tab, Resource, #state{protocol=grpc,
                             grpc_metadata=Metadata,
                             channel_pid=_ChannelPid}) ->
    %% ...
    %% Pass in along side `#{channel => ?MODULE}`?
    case opentelemetry_trace_service:export(Ctx, ExportRequest, #{channel => ?MODULE}) of
```

Would be great if anyone can point me to the right direction here

File: apps/opentelemetry_exporter/src/opentelemetry_exporter.erl (hunk 2 above)
Comment (tsloughter) [reply apps/opentelemetry_exporter/src/opentelemetry_exporter.erl:155]: Yea, that'd be good.

**Comment by kw7oe** [comment]
@tsloughter I'm trying to add the test for `grpc` but ending up have failing test on the second `grpc` test group due to `{error, no_endpoints}`. Test pass if I'm just having one `grpc` test group.

Do you have any idea what might caused this?

Here's my diff:

```diff
diff --git a/apps/opentelemetry_exporter/src/opentelemetry_exporter.erl b/apps/opentelemetry_exporter/src/opentelemetry_exporter.erl
index 79c89a9..59651e9 100644
--- a/apps/opentelemetry_exporter/src/opentelemetry_exporter.erl
+++ b/apps/opentelemetry_exporter/src/opentelemetry_exporter.erl
@@ -180,7 +180,11 @@ export(Tab, Resource, #state{protocol=grpc,
                              channel_pid=_ChannelPid}) ->
     ExportRequest = tab_to_proto(Tab, Resource),
     Ctx = grpcbox_metadata:append_to_outgoing_ctx(ctx:new(), Metadata),
-    case opentelemetry_trace_service:export(Ctx, ExportRequest, #{channel => ?MODULE}) of
+    Result = opentelemetry_trace_service:export(Ctx, ExportRequest, #{channel
+                                                                      =>
+                                                                      ?MODULE}),
+    ct:print("~p~n", [Result]),
+    case Result  of
         {ok, _Response, _ResponseMetadata} ->
             ok;
         {error, {Status, Message}, _} ->
diff --git a/apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl b/apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl
index d2f73ac..e8afaff 100644
--- a/apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl
+++ b/apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl
@@ -9,12 +9,12 @@
 -include_lib("opentelemetry/include/otel_span.hrl").
 
 all() ->
-    [{group, functional}, {group, http_protobuf}, {group,
-                                                   http_protobuf_gzip}, {group, grpc}].
+    [{group, functional}, {group, http_protobuf}, {group, http_protobuf_gzip}, {group, grpc}, {group, grpc_gzip}].
 
 groups() ->
     [{functional, [], [configuration, span_round_trip, ets_instrumentation_info]},
      {grpc, [], [verify_export]},
+     {grpc_gzip, [], [verify_export]},
      {http_protobuf, [], [verify_export]},
      {http_protobuf_gzip, [], [verify_export]}].
 
@@ -31,6 +31,9 @@ init_per_group(Group, Config) when Group =:= grpc ;
 init_per_group(http_protobuf_gzip, Config) ->
     application:ensure_all_started(opentelemetry_exporter),
     [{protocol, http_protobuf}, {compression, gzip} | Config];
+init_per_group(grpc_gzip, Config) ->
+    application:ensure_all_started(opentelemetry_exporter),
+    [{protocol, grpc}, {compression, gzip} | Config];
 init_per_group(_, _) ->
     application:load(opentelemetry_exporter),
     ok.
```

**Output:**

```
----------------------------------------------------
2021-11-14 20:34:26.825
{error,no_endpoints}
```

**Comment by kw7oe** [comment]
CI test failed due to collector:

```
ERROR: for opentelemetry-erlang_otel_1  Cannot start service otel: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "/otelcontribcol": permission denied: unknown

ERROR: for otel  Cannot start service otel: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "/otelcontribcol": permission denied: unknown
Encountered errors while bringing up the project.
Error: Process completed with exit code 1.
```

**Comment by tsloughter** [comment]
@kw7oe yea, I got that too and am not able to reproduce locally so haven't figured out how to debug it :(

**Comment by kw7oe** [comment]
> @kw7oe yea, I got that too and am not able to reproduce locally so haven't figured out how to debug it :(

This is referring to the CI/CD fail right? I think is a recent issues: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6238

I am able to replicate it using a different machine:
Screenshot 2021-11-16 at 10 14 07 AM

I guess our local machine is cached or something so we can't replicate it

File: apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl [hunk 3]
```diff
@@ -212,16 +221,22 @@ span_round_trip(_Config) ->
...
            end,
     {ok, State} = opentelemetry_exporter:init(#{protocol => Protocol,
+                                                compression => Compression,
                                                 endpoints => [{http, "localhost", Port, []}]}),
     Tid = ets:new(span_tab, [duplicate_bag, {keypos, #span.instrumentation_library}]),
 
+    %% Tempoararily adding this because without this, we would face
+    %% {error, no_endpoints} when attempt to export when we have more
+    %% than 1 gprc test case.
+    timer:sleep(500),
```
Comment (kw7oe) [inline apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl:239]: When we have more than one grpc test case, I'm getting `{error, no_endpoints}` error during `opentelemetry_trace_service:export`. It works fine if there's only one grpc test case running.

I'm not sure how to properly fix it, but having some delay in between seems to resolve that.

Comment (tsloughter) [reply apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl:239]: Hm, I had some theories until I saw where you put the sleep... It is after calling `init`. I guess it still could be somehow related to async shutdown/startup of grpcbox channels and endpoints, but I'm not sure my original idea of adding to `end_per_group` a wait for when everything in grpcbox has really stopped.

Comment (kw7oe) [reply apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl:239]: I see, I have also tried putting `timer:sleep(500)`:

- Before calling `init`
- In `end_per_group`  

but both doesn't resolve the issue.

Should I look further down into `grpcbox` side of things to check if is related to your theories?

Comment (tsloughter) [reply apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl:239]: If you have some time that'd be great. But I'd say we don't need to delay this PR to figure it out.

Comment (kw7oe) [reply apps/opentelemetry_exporter/test/opentelemetry_exporter_SUITE.erl:239]: Okay sure, I'll have a look into it when I have the time, will mark this as resolve soon.

**Review by tsloughter — APPROVED** [review]

**Pull request merged**